File: credit/credit.c
#include <stdio.h>
#include <cs50.h>

//Prototypes
long get_credit_number(void);
long get_product_sum(long num);
long last_digit(long num, long product);

int main(void)
{
    long c = get_credit_number();

    long p = get_product_sum(c);

    long digit = last_digit(c, p);

//Initialize Variables
    long n = 0;
    int count = 0;
    long first = 0;
    long second = 0;

//Get each indiviudal digit and count the digits
    while (c > 0)
    {
        n = c % 10;
        count++;
        if (c < 10)
        {
            first = n;
        }

        if (c > 10 && c < 100)
        {
            second = c % 100;
        }

        c = c / 10;
    }

//Determine credit card type based on conditions

//American Express
    if ((digit == 0 & count == 15 & second == 34) || digit == 0 & count == 15 & second == 37)
    {
        printf("AMEX\n");
    }

//MasterCard
    else if (digit == 0 & count == 16 & second > 50 & second <= 55)
    {
        printf("MASTERCARD\n");
    }

//VISA
    else if ((digit == 0 & count == 13 & first == 4) || (digit == 0 & count == 16 & first == 4))
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }

}

long get_credit_number(void)
{
    long c;
//Prompt user for input
    do
    {
        c = get_long("Number: ");
    }
    while (c <= 0);
    return c;
}

long get_product_sum(long num)
{
//Initialize Variables
    long digit = 0;
    int n = 0;
    long product = 0;
    long sum = 0;
//Get each indiviudal digit and count the digits
    while (num > 0)
    {
        digit = num % 10;
        n++;

//Multiply every other digit
        if (n % 2 == 0)
        {
            product = (digit * 2);
//Add the digits
            if (product > 9)
            {
                sum = sum + product % 10;
                product = product / 10;
            }
            sum = sum + product;
        }
        num = num / 10;
    }
    
    return sum;
}

long last_digit(long num, long product)
{
//Initialize Variables
    long digit = 0;
    int n = 0;
    int sum = product;
    long last;

//Get each indiviudal digit and count the digits
    while (num > 0)
    {
        digit = num % 10;
        n++;

//Add sum from previous function to other digits
        if (n % 2 != 0)
        {
            sum = sum + digit;
        }
        num = num / 10;
    }

    last = sum % 10;
    return last;
}

File: credit/credit.py
from cs50 import get_int, get_string


def main():
    # Get credit card number
    number = get_credit_number()
    # Determine if a valid credit card using Luhn’s algorithm
    product = get_product_sum(number)
    digit = last_digit(number, product)
    first, second = get_digits(number)
    # Print type of credit card or invalid
    print_credit_card(number, digit, first, second)


def get_credit_number():
    # Get positive credit number
    while True:
        num = get_int("Number: ")
        if num > 0:
            break
    return num


def get_product_sum(num):
    length = str(num)
    count = 0
    total = 0
    # Get each individual digit and count the digits
    for i in length:

        digit = int(num % 10)
        count += 1

        # Multiply every other digit
        if count % 2 == 0:

            product = digit * 2
            # Add the digits
            if product > 9:
                # Add the products and sum
                total += int(product % 10)
                product = int(product / 10)

            total += product

        num = num / 10

    return total


def last_digit(num, product):
    # Initialize Variables
    length = str(num)
    count = 0
    total = product

    # Get each individual digit and count the digits
    for i in length:

        digit = int(num % 10)
        count += 1

    # Add sum from previous function to other digits
        if count % 2 != 0:
            total = int(total + digit)

        num = num / 10

    last = int(total % 10)

    return last


def get_digits(num):
    # Change digits to a string and just obtain index to get values
    length = str(num)
    first = int(length[0])
    second = int(length[0] + length[1])
    return first, second


def print_credit_card(number, digit, first, second):
    num = str(number)
    length = len(num)

    # Validate credit card type
    if digit == 0:
        if length == 15 and second == 34 or length == 15 and second == 37:
            print("AMEX")
        elif length == 16 and second > 50 and second <= 55:
            print("MASTERCARD")
        elif length == 13 and first == 4 or length == 16 and first == 4:
            print("VISA")
        else:
            print("INVALID")

    else:
        print("INVALID")


main()
